### ISSUE TITLE
Silent exclusion of declaration files

### ISSUE THREAD

**Issue by jshado1** [Microsoft/TypeScript#24990]
## Search Terms

declaration files

## Suggestion

>Please note that the compiler does not include files that can be possible outputs; e.g. if the input includes index.ts, then index.d.ts and index.js are excluded. In general, having files that differ only in extension next to each other is not recomended.

This tiny note buried in the [docs](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html#details) cost me half a day of WTF. It also goes against official examples for declaration files, which tell the reader to do exactly what that note warns against. Ex [module-class.d.ts](https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-class-d-ts.html)

This is extremely confusing, especially since no indication is included in the output. Just silent failure.

At the very least, there should be a giant warning.

Also, WHY do you do this???

**Comment by mhegazy** [comment]
> It also goes against official examples for declaration files, which tell the reader to do exactly what that note warns against. Ex module-class.d.ts

not sure how is that the case.

> there should be a giant warning.

by default the compiler emits output in the same directory as the input. if this was a warring, every user by default will get a warning on the second compilation. do not think this is feasible.

**Comment by jshado1** [comment]
@mhegazy sorry, I missed the notification for your response.

>not sure how is that the case.

I don't know how to respond to that. Is the example I cited unclear, or are you saying you think what I described is not happening?

>by default the compiler emits output in the same directory as the input. if this was a warring, every user by default will get a warning on the second compilation. do not think this is feasible.

Ah, okay. I had not considered a case where same outDir as in (which would it makes sense). But, I think generating output into the same directory as the input is not a common use-case—in fact, I've never seen it: Every usage I've seen exports to either `/lib` or `/dist` (or some similar variation). Designing baffling behaviour for common usage in order to support an (albeit destructive) edge-case seems wrong.

Would it be possible to detect the case you described, and then ignore those files (maybe with a message saying they were ignored, if a verbose flag was passed), but in all other cases, not ignore them? I think the amount of complexity this would add to TypeScript would be relatively minor
